test_root/semgrep/celery-code-owner.yml

```diff
@@ -0,0 +1,104 @@
+      (from the `edx_django_utils.monitoring` module) in order for us
+      to correctly track code-owners for errors and in other monitoring.
+
+      For more information, see the Celery section of "Add Code_Owner
+      Custom Attributes to an IDA" in the Monitoring How-Tos of
+      <https://edx.readthedocs.io/projects/edx-django-utils>.
+    languages:
+      - python
+    patterns:
+      # Find functions with decorators containing the substring "task"
+      # in their name. This might end up with false positives, but
+      # there are a lot of variations on how we decorate Celery tasks.
+
+      # This pattern should match all decorators, whether or not
+      # they're called as a function (both `@foo(...)` and `@foo`)
+      # and whether or not there are other decorators above or below.
+      - pattern-either:
+        - pattern: |
+            @$TASK
+            def $F(...):
+              ...
+        - pattern: |
+            @$TASK(...)
+            def $F(...):
+              ...
+
+      # Restrict the decorators of interest to just ones with "task"
+      # in the name.
+      - metavariable-pattern:
+          metavariable: $TASK
+          patterns:
+            - pattern-regex: >-
+                [^\(]*task(\(|$)
+
+      # Filter out all of the properly annotated functions, leaving
+      # just the ones of interest.
+      - pattern-not: |
+          @set_code_owner_attribute
+          def $F(...):
+            ...
+      # This is an alternative approach that we have needed in rare cases.
+      - pattern-not: |
+          def $F(...):
+            ...
+            set_code_owner_attribute_from_module(...)
+
+    severity: WARNING
+
+  # This is like celery-missing-code-owner-function but for the `run`
+  # method of Task classes.
+  - id: celery-missing-code-owner-class
+    message: |
+      Celery task classes need to decorate their `run` method with
+      `@set_code_owner_attribute` (imported from `edx_django_utils.monitoring`)
+      in order for us to correctly track code-owners for errors and in other
+      monitoring. Alternatively, the `run` method can call
+      `set_code_owner_attribute_from_module`.
+
+      For more information, see the Celery section of "Add Code_Owner
+      Custom Attributes to an IDA" in the Monitoring How-Tos of
+      <https://edx.readthedocs.io/projects/edx-django-utils>.
+    languages:
+      - python
+    patterns:
+      - pattern: |
+          class $C(..., $SUPER, ...):
+            def run(...):
+              ...
+      - metavariable-pattern:
+          metavariable: $SUPER
+          patterns:
+            - pattern-regex: "Task$"
+
+      - pattern-not: |
+          class $C(..., $SUPER, ...):
+
+            @set_code_owner_attribute
+            def run(...):
+              ...
+
+      - pattern-not: |
+          class $C(..., $SUPER, ...):
+
+            @set_code_owner_attribute
+            def run(...):
```

Comment:
Author's note: I'm not actually sure if this works, or if you'd need to use Django's `method_decorator` here.

Reply:
I'm not sure exactly what this means. However, it did make me think about the fact that edx-lint has test for our various pylint features. Are you able to add that for this PR?

Reply:
1. Meaning, I'm not sure if you can actually put this decorator on a method. I don't have a specific reason to think it won't work, but there aren't any examples of people doing it, and I've seen a utility in Django that is apparently specifically for transforming function decorators into method decorators. I just don't know enough about scope and receivers in Python. Maybe it's actually fine!
2. I don't think there's a fast and simple way to add tests for this check. I think of it more like unit tests, which are more or less supposed to be self-validating.